expression/builtin_arithmetic.go

```diff
@@ -416,37 +416,48 @@ func (s *builtinArithmeticMinusIntSig) evalInt(row chunk.Row) (val int64, isNull
 		return 0, isNull, err
 	}
 	forceToSigned := s.ctx.GetSessionVars().SQLMode.HasNoUnsignedSubtractionMode()
-	isLHSUnsigned := !forceToSigned && mysql.HasUnsignedFlag(s.args[0].GetType().Flag)
-	isRHSUnsigned := !forceToSigned && mysql.HasUnsignedFlag(s.args[1].GetType().Flag)
+	isLHSUnsigned := mysql.HasUnsignedFlag(s.args[0].GetType().Flag)
+	isRHSUnsigned := mysql.HasUnsignedFlag(s.args[1].GetType().Flag)
 
-	if forceToSigned && mysql.HasUnsignedFlag(s.args[0].GetType().Flag) {
-		if a < 0 {
-			return 0, true, types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
+	switch {
+	case forceToSigned && isLHSUnsigned && isRHSUnsigned:
+		ua, ub, uMaxInt64, uMinInt64 := uint64(a), uint64(b), uint64(math.MaxInt64), uint64(math.MaxInt64+1)
+		if (ua >= ub && ua-ub > uMaxInt64) || (ua < ub && ub-ua > uMinInt64) {
+			return 0, true, types.ErrOverflow.GenWithStackByArgs("BIGINT", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
 		}
-	}
-	if forceToSigned && mysql.HasUnsignedFlag(s.args[1].GetType().Flag) {
-		if b < 0 {
-			return 0, true, types.ErrOverflow.GenWithStackByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
+	case forceToSigned && isLHSUnsigned && !isRHSUnsigned:
+		if (a < 0 && b > 0 && uint64(a) > uint64(math.MaxInt64+b)) ||
+			(a < 0 && b <= 0) ||
```

Comment:
what about 
``` sql
mysql> select cast(-9223372036854775809 as unsigned) - 0;
+--------------------------------------------+
| cast(-9223372036854775809 as unsigned) - 0 |
+--------------------------------------------+
|                                          0 |
+--------------------------------------------+
1 row in set, 1 warning (0.00 sec)
```

Reply:
Wrong result is caused by a bug in cast function, check #22827.

As for the query `select cast(-9223372036854775809 as unsigned) - 0`, if we run an equivalent query `select cast(9223372036854775808 as unsigned) - 0;`, the output is within expectation, leading to overflow.

```
select cast(9223372036854775808 as unsigned) - 0;                                                                                                             
(1690, "BIGINT value is out of range in '(9223372036854775808 - 0)'")
```

Reply:
It's my fault here.
The example above is the result of MySQL.
I test using this commit, TiDB returns the same result.